samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt

```diff
@@ -0,0 +1,514 @@
+                    )
+                }
+
+                // Set up symbol swatches
+                symbolSwatches[SnapRuleBehavior.RulesPreventSnapping] = createSwatch(rulesPreventSymbol)
+                symbolSwatches[SnapRuleBehavior.RulesLimitSnapping] = createSwatch(rulesLimitSymbol)
+                symbolSwatches[SnapRuleBehavior.None] = createSwatch(noneSymbol)
+            }.onFailure {
+                messageDialogVM.showMessageDialog(
+                    "Error loading geodatabase", it.message.toString()
+                )
+            }
+        }
+    }
+
+    /**
+     * Start the geometry editor to edit the geometry of the selected feature.
+     */
+    fun editFeatureGeometry() {
+        // Get the symbol for the selected feature
+        selectedFeature?.let { feature ->
+
+            // Use the symbol from the selected feature in the style of the geometry editor tool
+            val symbol =
+                (feature.featureTable as GeodatabaseFeatureTable).layerInfo?.drawingInfo?.renderer?.getSymbol(
+                    feature
+                )
+            geometryEditor.tool.style.apply {
+                vertexSymbol = symbol
+                feedbackVertexSymbol = symbol
+                selectedVertexSymbol = symbol
+                vertexTextSymbol = null
+            }
+
+            // Hide the selected feature
+            (feature.featureTable?.layer as? FeatureLayer)?.setFeatureVisible(feature, false)
+
+            // Start the geometry editor and center the map underneath the reticle
+            feature.geometry?.let { initialGeometry ->
+                viewModelScope.launch {
+                    mapViewProxy.setViewpointCenter(initialGeometry.extent.center)
+                }
+                geometryEditor.start(initialGeometry)
+                geometryEditor.selectVertex(0,0)
+            }
+        }
+    }
+
+    /**
+     * Stop the geometry editor and discard the changes made to the geometry.
+     */
+    fun discardGeometryChanges() {
+        // Discard the current edit
+        geometryEditor.stop()
+
+        // Reset the selection
+        resetSelections()
+    }
+
+    /**
+     * Stop the geometry editor, and update the previously identified feature with the new geometry.
+     */
+    fun saveGeometryChanges() {
+        // Stop the geometry editor and get the updated geometry
+        val finalGeometry = geometryEditor.stop()
+
+        // Update the feature with the new geometry
+        selectedFeature?.let { feature ->
+            feature.geometry = finalGeometry
+
+            viewModelScope.launch {
+                (feature.featureTable as GeodatabaseFeatureTable).updateFeature(feature)
+                    .onFailure { error ->
+                        messageDialogVM.showMessageDialog(
+                            "Error updating feature",
+                            error.message.toString()
+                        )
+                    }
+            }
```

Comment:
Could be simplified to use (with `as?`): 
```suggestion
                (feature.featureTable as? GeodatabaseFeatureTable)?.updateFeature(feature)
                    ?.onFailure { error -> messageDialogVM.showMessageDialog(error) }
```